packages/labs/analyzer/src/lib/javascript/modules.ts

```diff
@@ -5,77 +5,194 @@
  */
 
 import ts from 'typescript';
-import {Module, AnalyzerInterface, PackageInfo} from '../model.js';
+import {Module, AnalyzerInterface, PackageInfo, Declaration} from '../model.js';
 import {
   isLitElement,
   getLitElementDeclaration,
 } from '../lit-element/lit-element.js';
-import * as path from 'path';
 import {getClassDeclaration} from './classes.js';
 import {getVariableDeclarations} from './variables.js';
 import {AbsolutePath, absoluteToPackage} from '../paths.js';
 import {getPackageInfo} from './packages.js';
+import {DiagnosticsError} from '../errors.js';
+
+// Cache of Module models by path; invalidated when the sourceFile
+// or any of its dependencies change
+const moduleCache = new Map<AbsolutePath, Module>();
```

Comment:
This is where plain functions seem problematic for me. I'd ideally want the state kept with some kind of object so that collisions are contained. What happens if we analyze a program under different compiler settings? Do we have to force this `module.ts` module to be unloaded from the VM somehow?

Reply:
As discussed, will move the module cache onto the Analyzer class

Reply:
Done